src/org/labkey/test/tests/InvalidateSessionTest.java

```diff
@@ -0,0 +1,151 @@
+import org.openqa.selenium.Cookie;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.fail;
+import static org.labkey.test.WebTestHelper.getCookies;
+
+@Category({Daily.class})
+public class InvalidateSessionTest extends BaseWebDriverTest
+{
+    private static final String USER = "pwd_change_user@invalidatesessiontest.test";
+
+    @BeforeClass
+    public static void setupProject()
+    {
+        InvalidateSessionTest init = getCurrentTest();
+        init.doSetup();
+    }
+
+    @Override
+    protected @Nullable String getProjectName()
+    {
+        return null;
+    }
+
+    @Override
+    protected void doCleanup(boolean afterTest)
+    {
+        _userHelper.deleteUsers(afterTest, USER);
+    }
+
+    private void doSetup()
+    {
+        _userHelper.createUser(USER);
+
+        ApiPermissionsHelper permissionsHelper = new ApiPermissionsHelper(this);
+        permissionsHelper.addMemberToRole(USER, "Folder Administrator", PermissionsHelper.MemberType.user);
+        setInitialPassword(USER);
+    }
+
+    /*
+        Regression coverage for Secure Issue 51523: Invalidate sessions on password change
+     */
+    @Test
+    public void testSessionInvalidatesAfterPasswordChange() throws IOException
+    {
+        signOut();
+        signIn(USER);
+        Connection cn = createDefaultConnection();
+        SelectRowsResponse response;
+        SelectRowsCommand selectCmd = new SelectRowsCommand("auditLog", "UserAuditEvent");
+        try
+        {
+            response = selectCmd.execute(cn, "Home");
+            Assert.assertEquals("Did not establish the database connection before the password change", 200,
+                    response.getStatusCode());
+        }
+        catch (IOException | CommandException e)
+        {
+            throw new RuntimeException(e);
+        }
+
+        log("Changing the user password");
+        String newPassword = PasswordUtil.getPassword() + "&*&*";
+        goToMyAccount().clickChangePassword()
+                .setOldPassword(PasswordUtil.getPassword())
+                .setPassword1(newPassword)
+                .setPassword2(newPassword)
+                .clickSubmit();
+
+        log("Verifying the session is invalid");
+        try
+        {
+            selectCmd.execute(cn, getProjectName());
+            fail("Test should have thrown command exception");
+        }
+        catch (CommandException e)
+        {
+            Assert.assertEquals("Session should be expired", 401, e.getStatusCode());
+        }
+        signOut();
+    }
+
+    /*
+        Regression coverage for Secure Issue 31493: Test for session and cookie persistence through login and logout
+     */
+    @Test
+    public void testCookieAndSessionFromLogout() throws IOException
+    {
+        log("Capture the cookie after login");
+        Cookie beforeCookie = getCookies(getCurrentUser()).get(Connection.JSESSIONID);
```

Comment:
`WebTestHelper.getCookies` gets cached cookies. Use `getDriver().manage().getCookies()` to ensure that you're getting the live cookies from the browser.